### PR TITLE
Make it possible to build against on-chain dependencies

### PR DESCRIPTION
**Pull request by kklas**
## Description 

Context https://gist.github.com/kklas/d9edad7875548fae57b0c86bc8871fea

### Unresolved issues and discussion points:

**(1)** Due to how dependencies are resolved, some address mappings don't propagate properly through the graph when the same package exists as on-chain and source (transitive) dependency. This happens becuase the dependency graph treats packages of the same name and version as the same regradless of their location so sometimes the on-chain package ends up in the resulting `package_table` and since those don't have address names defined, it can cause undefined address name errors in the parent source packages. I alleviated this a bit by adding names for system packages ("std", "sui", "sui_system", and "deepbook") to the generated manifest in https://github.com/MystenLabs/sui/commit/2623a91fac278fd6a28bc7b24918faf264e11add but this doesn't solve the problem for all packages. It could be perhapse solved by keeping track of all the instances of the same package during dependency resolution and then resolving the addresses by collecting them from all. This is simple in principle but could get tricky to implement because overrides and address renamings also need to be respected. This could also be solved with a source discovery service where we could fetch the names from. In any case, the user can always do an override with a source package to go around this issue.

**(2)** There is now a conflict between old and new lockfiles. The old ones will use non-resolved names, so when building a package with the new binary and dependencies have lockfiles it will cause an incosistency. This is because lockfiles are only regenerated based on manifest and dependency manifest digests, which don't change here. This can be solved by bumping the lockfile version or forcing lockfile regeneration on resulting lockfile contents change https://github.com/MystenLabs/sui/blob/main/external-crates/move/tools/move-package/src/resolution/dependency_graph.rs#L240-L241

**(3)** Lockfiles don't store package's original name, so when loading dependency graph from lockfiles, error messages will print out resolved names instead of original which will be confusing to the user. I illustrated this issue in https://github.com/MystenLabs/sui/commit/b5c3093c4f7cedc6d0c58dc5398a10e73f8d72df. This can be solved by adding original package names to lockfiles.

**(4)** I've made only an in-memory cache for on-chain dependencies and each is fetched once per run for simplicity and seems fast enough. They can easilly be cached to disk in `~/.move` though with a bit of consideration about consistency. System packages can also be embedded in the binary to avoid RPC fetching.

**(5)** The `version` field can now be removed from all manifests in this repo. Building with the old binary will cause an error in that case. LMK if you want me to do that. Otherwise it will just cause an extra field warning when building with the new binary.

**(6)** To avoid confusion with variable names, the `custom_resolve_pkg_name` hook should perhaps be called `resolve_pkg_identifier` hook and `identifier` should be used in place of `pkg_name` and `resolved_name` etc. 

**(7)** `move-analyzer` doesn't work with both custom hooks and bytecode dependencies. I've managed to make it work with custom with a hacky solution where I've copied the whole crate to a new one and made it use the hooks (it required, among other things, to make it use tokio). But it still didn't work because it has some issues probably with interpreting the bytecode and while the server works it throws errors. I didn't dig deeper.

**(8)** `sui move test` doesn't work with on-chain packages. This is because of two separate reasons:
- First is that often the MoveStdlib package (0x1) ends up being included as an on-chain dependency which means that `#[test_only]` modules aren't included, but the test runner requires the `std::unit_test` module and fails with `Compilation in test mode requires passing the UnitTest module in the Move stdlib as a dependency`
- The second is due to the topo sort in `external-crates/move/tools/move-unit-test/src/test_runner.rs:83` that relies on `move-bytecode-utils/src/dependency_graph.rs` [stack trace](https://gist.github.com/kklas/51552100ffb4efbd265231cc79683339). Similar why publish was failing (and fixed in 6b28139fcddc0a28fdb8b5731ddc1b86acab5c98) it is because the compiler doesn't return the bytecode modules in the result and the graph construction fails due to missing dependency.

Both of these can fortunately be solved by doing an appropriate override, but the second issue means that we can't do unit tests when on-chain / bytecodes are present until this is resolved (it's very solveable though). The first issue can be solved by including the source for `0x1` instead of bytecode (and also probably for `0x2`, `0x3` and other framework packages) but there's some work to figure out how to embed the source code in the binary (and which version of it as there may be discrepencies between networks).

## Test Plan 

Unit and integration tests.

---

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: external-crates/move/tools/move-package/src/resolution/dependency_graph.rs [hunk 1]
```diff
@@ -128,7 +127,7 @@ impl PartialEq for Package {
         // comparing packages during insertion of externally resolved ones - an internally resolved
         // existing package in the graph would not be recognized as a potential different version of
         // the externally resolved one)
-        self.kind == other.kind && self.version == other.version
```
Comment (rvantonder) [inline external-crates/move/tools/move-package/src/resolution/dependency_graph.rs:131]: High-level comment on this commit: I'm a big fan of putting up this PR separately before everything else since it's clean / self-contained. We need to cover our bases with how this affects users / manifests that contain `version`. I.e., great if it's vestigial and we can conceptually remove it. To go along with that change, for example, we need to remove references to `version` in the sui repo in `examples` and elsewhere: https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/+file:Move%5C.toml+version+-repo:diem+-repo:aptos&patternType=standard&sm=1&groupBy=repo and communicate the change in the changelog

File: external-crates/move/tools/move-package/src/package_hooks.rs [hunk 2]
```diff
@@ -64,3 +70,14 @@ pub(crate) fn custom_package_info_fields() -> Vec<String> {
...
+pub(crate) fn custom_resolve_pkg_name(
+    pkg_path: &PathBuf,
+    manifest: &SourceManifest,
+) -> anyhow::Result<Symbol> {
+    if let Some(hooks) = &*HOOKS.lock().unwrap() {
+        hooks.custom_resolve_pkg_name(pkg_path, manifest)
+    } else {
+        Ok(manifest.package.name)
+    }
+}
```
Comment (rvantonder) [inline external-crates/move/tools/move-package/src/package_hooks.rs]: Seems reasonable!

File: external-crates/move/tools/move-package/src/resolution/dependency_graph.rs [hunk 3]
```diff
@@ -137,6 +137,9 @@ pub struct Dependency {
     pub subst: Option<PM::Substitution>,
     pub digest: Option<PM::PackageDigest>,
     pub dep_override: PM::DepOverride,
+    /// Original dependency name as defined in parent manifest since it can be different from the
+    /// resolved name. Used for printing user-friendly error messages.
+    pub dep_orig_name: PM::PackageName,
```
Comment (rvantonder) [inline external-crates/move/tools/move-package/src/resolution/dependency_graph.rs:158]: This commit looks like something great we can ship outright as preparation for compiling against on-chain dependencies, i.e., in separate PR. That would help reduce the footprint / size of the core feature and PR at the end!

**Review by rvantonder — COMMENTED** [review]
High-level: Thinking about how we can chop this out to make iterative progress, it looks like quite a few commits could be factored out and merged cleanly on their own as preparation for the core changes--especially the ones propagating info for error messages and code shuffling (54a97f361ef2d362ca092e099ea8a5d5b0289602). So, re: discussion points:

**(1)** & **(2)** I don't have good recommendations on these yet, will chew on it a bit more.

**(3)** I think this is a separable problem + solution that could work nicely, ando go in a separate PR for prep

**(4)** Seems reasonable--the default expectation for devs right now is to compile against source, so that is always an available fallback while we're working towards a v0 for compiling against on-chain deps. I.e., perf considerations can be a fast follow

**(5)** I think this is OK--we need to take care of removing references from docs / code in the Sui repo and also communicate the change in Changelogs, etc. though. Emitting a warning instead of a hard error in a next release would be a bit more user-friendly, followed by a hard error in a subsequent release.

**(6)** Sounds good--these renames seem like another set of candidate changes that would be great to separate and merge before the core functional changes.

To summarize, I wonder if we can create separate PR(s) for the mechanical changes required, with the core changes following up. My sense is that might make it easier to think / reason about the core code changes and dealing with **(1)** and **(2)**, and also to give room for input from @amnn.

**Comment by kklas** [comment]
> High-level: Thinking about how we can chop this out to make iterative progress

It should be possible to merge each of these commits on their own. A cool thing is that everything is fully backwards compatible until custom Sui hooks get implemented.

The downside of splitting up the PR is that there is some overhead in getting indiviual PRs to be mergable, but I can do that if you think this will make the overall process easier. Here's how I think commits can be groupped up into separate PRs:

1) 
[cherry-pick][move] enable bytecode as dependencies

2) 
[move-package] remove version from manifest package section and dependency
[move-package] fix tests for version removal

3)
[move-package] fix docstring ascii being parsed as doctest

4)
[move-package] add custom_resolve_pkg_name hook
[on-chain-deps] add tests for custom_resolve_pkg_name hook

5)
[on-chain-deps] use original package names in resolution error messages
[on-chain-deps] use original root pkg name in dependency graph error messages
[on-chain-deps] use original root pkg name in resolution graph error messages
[on-chain-deps] add a test for pkg name resolution conflict from lockfiles

6) 
[on-chain-deps] collect overrides from collect_graphs

7)
[on-chain-deps] add resolve_version hook
[on-chain-deps] add tests for version resolution
[on-chain-deps] use versions in dimanod_problem_dep_transitive_nested_override test

8) 
[on-chain-deps][refactor] store orig pkg name with DependencyGraphInfo to simplify function calls

9)
[on-chain-deps] change Custom dependency to OnChain

10)
[on-chain-deps] implement sui package hooks
[on-chain-deps] add addresses for system packages to generated manifest
[on-chain-deps] exclude on-chain dependencies from source verification


These can be amended to some existing commits:
[on-chain-deps] remove pkg_path param from name and version hooks
[on-chain-deps] run cargo format

File: external-crates/move/tools/move-package/src/compilation/compiled_package.rs [hunk 4]
```diff
@@ -575,11 +577,27 @@ impl CompiledPackage {
...
+        // If bytecode dependency is not empty, do not allow renaming
+        if !bytecode_deps.is_empty() {
+            if let Some(pkg_name) = resolution_graph.contains_renaming() {
+                anyhow::bail!(
+                    "Found address renaming in package '{}' when \
+                    building with bytecode dependencies -- this is currently not supported",
+                    pkg_name
+                )
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kklas) [reply external-crates/move/tools/move-package/src/compilation/compiled_package.rs:594]: This commit in particular I have cherry picked from upstream so I'm also not super familiar with details here. I'd rather leave it as is and perhaps do this change some other time.

File: external-crates/move/tools/move-package/src/resolution/dependency_graph.rs (hunk 1 above)
Comment (kklas) [reply external-crates/move/tools/move-package/src/resolution/dependency_graph.rs:131]: > To go along with that change, for example, we need to remove references to version in the sui repo in examples and elsewhere

As far as I'm concerned we can do that but we need to consider is that old versions of the tool will then error when building packages against master because version field is mandatory.

File: external-crates/move/tools/move-package/src/resolution/dependency_graph.rs (hunk 3 above)
Comment (kklas) [reply external-crates/move/tools/move-package/src/resolution/dependency_graph.rs:158]: This issue only happens when `custom_resolve_pkg_name` hook is used and returns something that's different than the manifest so it might not make sense to ship it before the hook code (that PR would be confusing).

**Comment by kklas** [comment]
I've added unresolved issue / discussion point **(7)**.

**Comment by rvantonder** [comment]
Thanks for your patience! 

> The downside of splitting up the PR is that there is some overhead in getting indiviual PRs to be mergable, but I can do that if you think this will make the overall process easier.

So I'd love if we could tackle this bit-by-bit and make it a bit more digestible for review; I do think we'll need to do that to make progress here (reviewing it all is a bit intimidating for me to review personally 😅) and the codebase can shift and cause conflicts (which looks like already happened).

> A cool thing is that everything is fully backwards compatible until custom Sui hooks get implemented. [... some consideration be taken about the rollout (because e.g. of (2), (3), (5)) ]

That is great. I think it's reasonable to bundle if it doesn't have user-visible impact and turns out to be relatively small (less than 1KLOC change). Considering the user-visible impact of (2), that might not be the case. Side question: do we strictly need (2) to enable on-chain dependencies? If not, we might consider punting on this so it's not blocking the actual functionality.

>  But I think everything should probably be commited within the same version cutoff

I agree with the convenience but it might be a bit difficult to pull off in practice, I think we have some options here:

- guard the new functionality behind a flag, allowing incremental implementation / iteration without affecting functionality until it's complete (my personal preference). Overhead is including the flag, but it makes things manageable without burden / pressure to review for a particular version cutoff.
- structure review with separate PRs that build on each other, with review / approval allowing to merge PRs into a base PR until it's all ready, and a final approval of the base PR. This will need some overhead to set up, and also to avoid making intermediate changes between PRs, because that will probably require a lot of rebasing on your part. It will help make review manageable though.

I'm not very confident about a big PR with separate commits, because it makes review very difficult to isolate and incorporate changes that come with feedback.

I'm also curious, to what extent you think some of the discussion points could be helped by having a feature flag on the implementation. E.g., to explicitly warn on known limitations.

Let me know your thoughts!

**Comment by rvantonder** [comment]
Thanks for the detailed breakdown, that's really helpful and clear! About resolving outstanding issues, like `move-analyzer` choking--that's tricky! For this, and maybe other issues that cause friction, it seems like there are two routes: (a) go deeper down the rabbit hole to resolve things (like with `move-analyzer`) (b) punt for and accept that compiling with on-chain dependencies might lead to a poor user experience here (we can explicitly list known limitation of compiling against on-chain dependencies, and mark it as an experimental mode for compiling). Right now, I don't think we have dedicated resources to devote to resolving issues in `move-analyzer` ourselves, but maybe in the future 🥲 . So I'm not opposed to (b), for example, as long as we're upfront about the limitations / state of implementation. 

> I'm not a fan of this approach because you still then need to merge the base PR into main which will require another look / review / rebase.

Yep, makes sense!

**Comment by kklas** [comment]
> So I'm not opposed to (b), for example, as long as we're upfront about the limitations / state of implementation.

I agree with this. Something with bad UX is still better than nothing. I'm actually using the toolchain from this PR to build a package with on-chain deps, and although it's a bit inconvenient without `move-analyzer` it's alright.

Anyway, I have found another issue marked with **(8)**. I will look into it further when I find some time.

**Comment by rvantonder** [comment]
> Anyway, I have found another issue marked with (8).

Gotcha, this one seems like a bit of a bummer. In an event, I think we're agreed on "on-chain deps support in an imperfect state is better than nothing" and as long as the code path is relatively clean for supporting that, I think it's reasonable to go ahead with PRs that start chopping up the feature for review.

**Comment by kklas** [comment]
@rvantonder @amnn so for adding package names and root resolved `id` and `version` to the lockfile, I had something like this in mind:

image

In summary we:
- add a new `root` field at the top level of lockfile which will contain the resolved `id` and `version` of the root package
  - this is needed just for the external resolver and will be leveraged [here](https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L490-L503) to load the corract thata into the graph (for non-external we resolve them via hooks with manifest)
- in the `dependencies` field of each package (and root):
  - the `name` field will be renamed to `id` because that's what effectively is since after the id resolution hook has landed
  - another field is added called `name` which will store the original package name (as defined in its manifest) for error messages (a note on this below [1])
    - this will be used [here](https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L1080) and other similar places in `read_from_lock` to load the correct data into the graph (edges)
- in each package section we rename the `name` field to `id` that uniquely identifies a package

[1] Now a question may arise: "why are package names stored in the dependencies section of each parent package instead of the package itself?". The reason for that is that package names are, in the graph, stored in edges betwee the nodes (packages). It was done this way in the "id resolution hook" PR because it's a more natural solution implementation wise (storing them in the package would be possible in principle but a way more intrusive implementation), and also because in principle it is the more correct solution. To illustrate this I will give an example.

Suppose we're resolving a graph with both a source and on-chain transitive dependency to `Foo` package. In case of the source dependency we can defenitively load its name from the manifest as `Foo` (and there is a consistency check for that in `resolved_graph.rs` [here](https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-package/src/resolution/resolution_graph.rs#L115-L137)), but when the same dependency is loaded from an on-chain reference, its name doesn't exist (on-chain) and there's nothing we can check against. So in case of on-chain dependencies we allow for arbitrary names in the parent manifest, and store this name in the edge in the graph (`Dependency` struct [here](https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L160)) instead of the node (`Package` struct [here](https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L125-L131)) so that this info is retained for constructing correct error messages (and possibly other things in the future).

Therefore, when constructing the lockfile, it is also more natural store the dependency names in the edges (`dependencies` list of each packages) instead of the package section, and retain this information so that the graph can be faithfully reconstructed from the lockfile.

**Comment by rvantonder** [comment]
Thanks @kklas, I quickly went over that and sounds reasonable to me, in general! I'll go over it a bit more deeply considering our plans to add more to the lock file here. My overarching comment is that we want to namespace resolution by network (chain identifier), so there will be a multiplicity of move packages we track in the lock file going forward.

**Comment by rvantonder** [comment]
Update this side @kklas: we're going over some designs for the `Move.lock` around tracking addresses and this overlaps a bit. Will report back once we have more clarity from our internal discussions, soon!
